fixes buffer count and window count

frankpepermans opened the pull request:
This fix should properly handle `buffer` on `count` and `window` on `count`, when using the `startBufferEvery` parameter.

This *is* a breaking change:
- startBufferEvery should now work correctly
- when the parent `Stream` completes, we no longer add the remainder `buffer` as a final `event`:
i.e. if the `buffer` is set to group by 3 events, and the parent `Stream` outputs `1, 2, 3, 4`, then the only `event` on `buffer(3)` would be `[1, 2, 3]`. 
Before, we would also dispatch `[4]` as the remainder `onDone`.

frankpepermans commented:
argh, this broke a few `Window` tests because I've changed the `onDone` handling, will check...

frankpepermans commented:
fixed, window was calling `onDone` too soon, which also closed the `Sampler` too soon, causing the last event to not emit in some tests

VerTiGoEtrex commented on lib/src/samplers/buffer_strategy.dart:
```diff
@@ -196,31 +194,30 @@ class _OnCountSampler<T, S> extends StreamView<S> {
       throw new ArgumentError('count cannot be null');
-    } else if (skip > count) {
-      throw new ArgumentError('skip cannot be greater than count');
     }
```
Not sure of your philosophy on bad input checking, but maybe add a check for `count < 0` -- currently negatives will throw deep inside the buffer handler code.

VerTiGoEtrex commented on test/transformers/buffer_count_test.dart:
```diff
@@ -35,15 +35,14 @@ void main() {
+    }, count: expectedOutput.length));
   });
 
   test('rx.Observable.bufferCount.skip', () async {
```
Maybe update the names of these tests to reflect the new behavior/param name. Or add a "todo" ;)

frankpepermans commented:
Good call :)

Added checks:
count needs to be `>= 1`
startBufferEvery needs to be `>= 0` (if `null`, then it defaults to 0)

renamed some tests, added the above checks in tests.

Finally, added a new op, `pairwise`, which is a shorthand for `bufferCount(2, 1)`, see https://github.com/ReactiveX/rxdart/issues/205

frankpepermans commented:
Yup that's why it's commented out,
I was following the docs where it stated:

"Collects values from the past as an array, and emits that array only when its size reaches bufferSize"

But I'm totally fine with allowing the final value in again.

Just need to doublecheck if pairwise has the same logic behind :)

frankpepermans commented:
@brianegan added the remaining buffer (if any) again when `onDone` triggers

brianegan commented on lib/src/observables/observable.dart:
```diff
@@ -1201,43 +1201,45 @@ class Observable<T> extends Stream<T> {
   ///
-  /// If [skip] is provided, each group will start where the previous group
-  /// ended minus the [skip] value.
+  /// If [startBufferEvery] is provided, each group will start where the previous group
```
I feel like this comment is just slightly difficult to understand / maybe a bit wrong with the new implementation? The RxJS docs use this description:

  > Interval at which to start a new buffer. For example if startBufferEvery is 2, then a new buffer will be started on every other value from the source. A new buffer is started at the beginning of the source by default.

frankpepermans replied:
Yup will update

brianegan commented on lib/src/observables/observable.dart:
```diff
@@ -1201,43 +1201,45 @@ class Observable<T> extends Stream<T> {
   ///
-  /// If [skip] is provided, each group will start where the previous group
-  /// ended minus the [skip] value.
+  /// If [startBufferEvery] is provided, each group will start where the previous group
```
Might need the same wording update as mentioned above

brianegan commented on lib/src/observables/observable.dart:
```diff
@@ -2291,23 +2304,24 @@ class Observable<T> extends Stream<T> {
   ///
-  /// If [skip] is provided, each group will start where the previous group
-  /// ended minus the [skip] value.
+  /// If [startBufferEvery] is provided, each group will start where the previous group
```
Wording

brianegan commented on lib/src/samplers/buffer_strategy.dart:
```diff
@@ -196,31 +194,35 @@ class _OnCountSampler<T, S> extends StreamView<S> {
       throw new ArgumentError('count cannot be null');
-    } else if (skip > count) {
-      throw new ArgumentError('skip cannot be greater than count');
+    } else if (count < 1) {
```
Should we consider using `assert` here instead of throwing errors? That way these won't impact perf in production apps?

frankpepermans replied:
Not sure, I suspect in most cases these args will be constant values, i.e. bufferCount(2, 1).

However should you write it this way: bufferCount(i, j) which could be anything, then your app might still fail in production.

It's only a null and an int compare check really.

Side note, I was building a benchmark test branch for all our code, so maybe we can revisit this later?

frankpepermans replied:
So we keep the checks I assume?

brianegan replied:
Yep, let's keep em for now!

brianegan commented on test/transformers/pairwise_test.dart:
```diff
@@ -0,0 +1,48 @@
+import 'package:test/test.dart';
+
+void main() {
+  test('rx.Observable.pairwise', () async {
```
I wouldn't adjust these tests, but in the future I think we should use Stream Matchers from the test package a bit more instead of these manual counts / expectAsync calls.

frankpepermans replied:
Ah, lazy copy pasting of other tests, agreed ;)

frankpepermans commented:
Fixes `pairwise`, the diff to `bufferCount(2, 1)` here is, that `pairwise `does *not* emit the remaining buffer `onDone`.

e.g.
```
Observable.range(1, 4)
    .bufferCount(2, 1)
    .listen(print); // [1, 2], [2, 3], [3, 4], [4]
```

```
Observable.range(1, 4)
    .pairwise()
    .listen(print); // [1, 2], [2, 3], [3, 4]
```
